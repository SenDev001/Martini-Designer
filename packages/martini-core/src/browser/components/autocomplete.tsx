import * as React from "react";
import styled from "styled-components";
import { ElementExt } from '@phosphor/domutils';

export interface Suggestion {
    label: string;
    iconClass?: string;
}

export interface AutocompleteProps {
    value: string;
    suggestions: Suggestion[];
    onInputChange: (input: string) => void;
    onInputBlur?: (e: React.FocusEvent<HTMLInputElement>, input: string) => void;
    onSuggestionSelect?: (input: string) => void;
    onInputKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
    onInputKeyUp?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
    inputStyle?: React.CSSProperties;
    autoFocus?: boolean;
}

const UlStyles = styled.ul`
    color: var(--theia-input-foreground);
    background-color: var(--theia-input-background);
    border: 1px solid var(--theia-inputOption-activeBorder);
    border-top: 0;
    position: absolute;
    top: 25px;
    left: 0px;
    width: calc(100% - 2px);
    list-style: none;
    margin-top: 0;
    max-height: 143px;
    overflow-y: auto;
    padding-left: 0;
    z-index: 20;

    li {
        padding: 4px;
    }

    .suggestion-active, li:hover {
        background-color: var(--theia-list-activeSelectionBackground);
        color: var(--theia-input-foreground);
        cursor: pointer;
    }
`;

export const Autocomplete: React.FC<AutocompleteProps> = ({
    value,
    suggestions,
    onInputChange,
    onInputBlur,
    onSuggestionSelect,
    onInputKeyDown,
    onInputKeyUp,
    inputStyle,
    autoFocus
}) => {
    const [activeSuggestion, setActiveSuggestion] = React.useState(0);
    const [filteredSuggestions, setFilteredSuggestions] = React.useState([] as Suggestion[]);
    const [showSuggestions, setShowSuggestions] = React.useState(false);
    const mounted = React.useRef(true);
    React.useEffect(() => {
        if (autoFocus && inputRef.current)
            inputRef.current.setSelectionRange(0, value.length);
        return () => { mounted.current = false; };
    }, []);
    const inputRef = React.useRef<HTMLInputElement>(null);
    const ulRef = React.useRef<HTMLUListElement>(null);

    const filter = (input: string) => input.length === 0 ? suggestions : suggestions.filter(
        suggestion =>
            suggestion.label.toLowerCase().includes(input)
    );

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const input = e.target.value;
        setActiveSuggestion(0);
        setFilteredSuggestions(filter(input.toLowerCase()));
        setShowSuggestions(true);
        onInputChange(input);
    };

    const onClick = (suggestion: Suggestion) => {
        setActiveSuggestion(0);
        setFilteredSuggestions([]);
        setShowSuggestions(false);
        onInputChange(suggestion.label);
        if (onSuggestionSelect)
            onSuggestionSelect(suggestion.label);
        inputRef.current?.focus();
    };

    const onKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === "Enter" && showSuggestions && filteredSuggestions.length !== 0) {
            const label = filteredSuggestions[activeSuggestion].label;
            setActiveSuggestion(0);
            setShowSuggestions(false);
            onInputChange(label);
            if (onSuggestionSelect)
                onSuggestionSelect(label);
            return;
        }
        else if (e.key === "Escape" && showSuggestions) {
            setShowSuggestions(false);
            return;
        }
        else if (e.key === "ArrowUp") {
            e.preventDefault();
            let next = activeSuggestion - 1;

            if (next < 0)
                next = filteredSuggestions.length - 1;

            setActiveSuggestion(next);
            if (ulRef.current)
                ElementExt.scrollIntoViewIfNeeded(ulRef.current, ulRef.current.childNodes[next] as HTMLLIElement);
            return;
        }
        else if (e.key === "ArrowDown") {
            e.preventDefault();
            let next = activeSuggestion + 1;

            if (next >= filteredSuggestions.length)
                next = 0;

            setActiveSuggestion(next);
            if (ulRef.current)
                ElementExt.scrollIntoViewIfNeeded(ulRef.current, ulRef.current.childNodes[next] as HTMLLIElement);
            return;
        }
        else if (!showSuggestions && e.key === " " && e.ctrlKey) {
            setShowSuggestions(true);
            return;
        }

        if (onInputKeyUp)
            onInputKeyUp(e);
    };

    const onBlur = (e: React.FocusEvent<HTMLInputElement>) => {
        // when e.relatedTarget === null it means it probably click one of the suggestions
        if (e.relatedTarget !== null && onInputBlur) {
            onInputBlur(e, value);
            setShowSuggestions(false);
        }
    };

    let suggestionsListComponent;

    if (showSuggestions && filteredSuggestions.length) {
        suggestionsListComponent = (
            <UlStyles ref={ulRef}>
                {filteredSuggestions.map((suggestion, index) => {
                    return (
                        <li
                            className={index === activeSuggestion ? "suggestion-active" : undefined}
                            key={suggestion.label}
                            onClick={() => onClick(suggestion)}
                        >
                            {suggestion.label}
                        </li>
                    );
                })}
            </UlStyles>
        );
    }

    return <>
        <input
            type="text"
            autoFocus={autoFocus}
            ref={inputRef}
            onChange={onChange}
            onBlur={onBlur}
            value={value}
            onKeyDown={onInputKeyDown}
            onKeyUp={onKeyUp}
            className="theia-input"
            style={inputStyle}
        />
        {suggestionsListComponent}
    </>;
};
