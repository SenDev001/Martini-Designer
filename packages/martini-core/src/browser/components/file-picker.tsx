import * as React from "react";

export interface FilePickerProps {
    extensions?: string;
    placeholder?: string;
    value?: File | string;
    onChange?: (fileOrUrl: File | string) => void;
}

export const FilePicker: React.FC<FilePickerProps> = ({extensions, placeholder, value, onChange, children}) => {
    const [file, setFile] = React.useState(!value ? "" : value instanceof File ? value.name : value);
    const fileInput = React.useRef<HTMLInputElement>(null);
    const [invalidDrop, setInvalidDrop] = React.useState(false);

    const isValidFile = (file: File) => {
        if (!extensions)
            return true;
        return extensions.split(/,/).some(ext => file.name.endsWith(ext));
    };
    const handleFiles = (files: FileList) => {
        if (files && files.length !== 0) {
            const first = files.item(0);
            if (first && isValidFile(first)) {
                setFile(first.name);
                if (onChange)
                    onChange(first);
                return true;
            }
        }
    };
    const handleFileInputChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        const files = e.target.files;
        if (files)
            handleFiles(files);
    };
    const handleTextInputChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        setFile(e.target.value);
        onChange!(e.target.value);
    };

    const handleDrop = (e: React.DragEvent) => {
        if (!handleFiles(e.dataTransfer.files)) {
            e.dataTransfer.dropEffect = "none";
            e.dataTransfer.effectAllowed = "none";
            setInvalidDrop(true);
            setTimeout(() => setInvalidDrop(false), 3000);
        }
    };

    return <div
        style={{display: "grid", gridTemplateColumns: "1fr max-content"}}
        onDrop={handleDrop}>
        <input type="file"
               style={{display: "none"}}
               ref={fileInput}
               accept={extensions}
               onChange={handleFileInputChanged}/>
        <input type="text"
               className="theia-input"
               placeholder={placeholder}
               onChange={handleTextInputChanged}
               style={{border: !invalidDrop ? "" : "var(--theia-border-width) solid var(--theia-errorForeground)"}}
               value={file}/>
        <input type="button"
               className="theia-button"
               value="Browse..."
               onClick={() => fileInput.current && fileInput.current.click()}/>
        {children}
    </div>;
};
