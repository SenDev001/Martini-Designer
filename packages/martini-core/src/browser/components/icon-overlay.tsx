import * as React from "react";

export interface IconProps {
    iconClass: string;
}

export const Icon: React.FC<IconProps> = ({ iconClass, children }) => (<div
    className={`martini-tree-icon ${iconClass}`}
    style={{ position: "relative", marginRight: "4px" }}
>
    {children}
</div>);

export interface OverlayIconProps extends IconProps {
    position: "top-left" | "top-right" | "bottom-left" | "bottom-right";
    width?: number;
    height?: number;
}

export const OverlayIcon: React.FC<OverlayIconProps> = (props) => {
    return <div className={props.iconClass} style={getStyle(props)} />;
};

const getStyle = (props: OverlayIconProps) => {
    const width = props.width === undefined ? 8 : props.width;
    const height = props.height === undefined ? 8 : props.height;
    const style: any = {
        backgroundSize: `${width}px ${height}px`,
        backgroundPosition: "center center",
        backgroundRepeat: "no-repeat",
        position: "absolute",
        width,
        height
    };

    switch (props.position) {
        case "top-left":
            style.left = "0";
            style.top = "0";
            break;
        case "top-right":
            style.right = "0";
            style.top = "0";
            break;
        case "bottom-left":
            style.left = "0";
            style.bottom = "0";
            break;
        case "bottom-right":
            style.right = "0";
            style.bottom = "0";
            break;
    }
    return style;
};
