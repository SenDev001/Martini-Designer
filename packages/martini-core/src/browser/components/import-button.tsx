import { CommandRegistry } from "@phosphor/commands";
import { Menu } from "@phosphor/widgets";
import { SingleTextInputDialog } from "@theia/core/lib/browser";
import * as React from "react";
import { isValidUrl } from "../../common/fs/file-util";

export interface ImportButtonProps {
    onFileImport?: (file: File) => void;
    onUrlImport?: (url: string) => void;
    extensions?: string;
}

export const ImportButton: React.FC<ImportButtonProps> = ({ onFileImport, onUrlImport, extensions }) => {
    const fileInputRef = React.useRef<HTMLInputElement>(null);
    const handleClick = (e: React.MouseEvent<HTMLInputElement>) => {
        const commands = new CommandRegistry();
        if (onFileImport) {
            commands.addCommand("fromFile", {
                label: "From File",
                execute: () => {
                    fileInputRef.current?.click();
                }
            });
        }

        if (onUrlImport) {
            commands.addCommand("fromUrl", {
                label: "From URL",
                execute: async () => {
                    const dlg = new SingleTextInputDialog({
                        title: "Enter a URL",
                        confirmButtonLabel: "Import",
                        validate: (url: string) => {
                            if (!isValidUrl(url))
                                return "Not a valid URL.";
                            return true;
                        }
                    });
                    const result = await dlg.open();
                    if (result) onUrlImport(result);
                }
            });
        }
        const menu = new Menu({
            commands
        });
        if (onFileImport)
            menu.addItem({
                command: "fromFile"
            });
        if (onUrlImport)
            menu.addItem({
                command: "fromUrl"
            });
        const bounds = e.currentTarget.getBoundingClientRect();
        menu.open(bounds.left, bounds.bottom);
    };

    const handleFileInput = (e: React.ChangeEvent<HTMLInputElement>) => {
        const files = e.target.files;
        if (files && files.length !== 0) {
            const first = files.item(0);
            if (first) onFileImport!(first);
        }
    };

    return <div>
        <input
            type="file"
            ref={fileInputRef}
            style={{ display: "none" }}
            accept={extensions}
            onChange={handleFileInput}
        />
        <input
            type="button"
            className="theia-button"
            style={{ marginLeft: 0 }}
            value="Import..."
            onClick={handleClick} />
    </div>;
};
