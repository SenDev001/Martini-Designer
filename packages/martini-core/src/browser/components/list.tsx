import messages from "martini-messages/lib/messages";
import * as React from "react";
import styled from "styled-components";
import { ElementExt } from '@phosphor/domutils';

export interface ListItem {
    label: string;
    data?: any;
    tooltip?: string;
    iconClass?: string;
    selected?: boolean;
}

export interface ListProps {
    items: ListItem[];
    onSelectionChanged?: (item: ListItem) => void;
    onDoubleClick?: (item: ListItem) => void;
    onTextChange?: (text: string) => void;
    filtered?: boolean;
    style?: React.CSSProperties;
    focus?: boolean;
    searchPlaceholder?: string;
}

const ListStyles = styled.ul`
    padding: 0px;
    margin-top: 0px;
    width: 100%;
    height: 100%;
    overflow-y: auto;

    .item {
        list-style: none;
        padding: 4px;
        display: grid;
        grid-template-columns: max-content 1fr;
        cursor: pointer;
    }

    .message {
        list-style: none;
        padding: 4px;
        display: grid;
        justify-items: center;
    }

    .item-selected {
        background-color: var(--theia-list-activeSelectionBackground);
    }
`;

export const List: React.FC<ListProps> = (props) => {
    const [items, setItems] = React.useState(props.items);
    const searchInputRef = React.useRef<HTMLInputElement>(null);
    const listRef = React.useRef<HTMLUListElement>(null);

    const doFocus = () => {
        if (props.focus) {
            if (props.filtered && searchInputRef.current)
                searchInputRef.current.focus();
            else if (listRef.current)
                listRef.current.focus();
        }
    };

    React.useEffect(doFocus, []);
    React.useEffect(doFocus, [props.focus]);
    React.useEffect(() => setItems(props.items), [props.items]);

    const handleClick = (item: ListItem) => {
        const updatedItems = items.map(item => ({ ...item }));
        updatedItems.forEach(other => other.selected = other.label === item.label);
        setItems(updatedItems);
        if (props.onSelectionChanged)
            props.onSelectionChanged(item);
    };

    const handleDoubleClick = (item: ListItem) => {
        if (props.onDoubleClick)
            props.onDoubleClick(item);
    };

    const handleKeyDown = (e: React.KeyboardEvent) => {
        if (e.key === "ArrowUp" || e.key === "ArrowDown") {
            if (items.length === 0)
                return;
            e.preventDefault();
            let index = items.findIndex(other => other.selected);
            if (e.key === "ArrowUp")
                index = index - 1 < 0 ? items.length - 1 : index - 1;
            else if (e.key === "ArrowDown")
                index = index + 1 >= items.length ? 0 : index + 1;

            if (listRef.current) {
                const liElem = listRef.current.children.item(index);
                if (liElem)
                    ElementExt.scrollIntoViewIfNeeded(listRef.current, liElem);
            }
            handleClick(items[index]);
        }
    };

    const list = <ListStyles
        tabIndex={0}
        className="theia-input"
        onKeyDown={e => handleKeyDown(e)}
        style={!props.filtered ? props.style : undefined}
        ref={listRef}
    >
        {props.filtered && items.length === 0 ?
            <li className="message">{messages.no_matches}</li> :
            items.map((item, i) => (
                <li
                    key={i}
                    onClick={() => handleClick(item)}
                    onDoubleClick={() => handleDoubleClick(item)}
                    className={"item " + (item.selected ? "item-selected" : "")}
                    title={item.tooltip}
                >
                    {item.iconClass && <span className={item.iconClass} />}
                    {item.label}
                </li>
            ))}
    </ListStyles>;

    if (!props.filtered)
        return list;

    const handleSearch = (search: string) => {
        if (props.onTextChange)
            props.onTextChange(search);
        else {
            search = search.toLowerCase();
            let _items;
            if (!search || search.length === 0)
                _items = [...props.items];
            else
                _items = props.items.filter(item => item.label.toLowerCase().includes(search));
            _items.forEach((item, i) => item.selected = i === 0);
            setItems(_items);
            if (props.onSelectionChanged && _items.length !== 0)
                props.onSelectionChanged(_items[0]);
        }
    };

    return <div style={{
        display: "grid",
        gridTemplateColumns: "1fr",
        gridRowGap: "4px",
        gridTemplateRows: "max-content 1fr",
        ...props.style
    }}>
        <input
            ref={searchInputRef}
            type="text"
            className="theia-input"
            onChange={e => handleSearch(e.target.value)}
            onKeyDown={handleKeyDown}
            placeholder={props.searchPlaceholder || messages.search_placeholder}
        />
        {list}
    </div>;
};
