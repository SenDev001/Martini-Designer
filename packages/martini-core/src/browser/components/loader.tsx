import * as React from "react";
import styled from "styled-components";

const LoaderStyles = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;

    .wrapper {
        display: grid;
        width: 100%;
        height: 100%;
        justify-items: center;
        grid-row-gap: 10px;
    }

    .progress-loader-end {
        align-self: end;
    }

    .progress-loader-center {
        align-self: center;
    }

    .progress-message {
        align-self: start;
    }
`;

export interface LoaderProps {
    message?: string;
    style?: React.CSSProperties;
}

export const Loader: React.FC<LoaderProps> = ({ message, style }) => (<LoaderStyles style={style}>
    <div className="wrapper">
        <div className={message ? "loader-large progress-loader-end" : "loader-large progress-loader-center"} />
        {message && <div className="progress-message">{message}</div>}
    </div>
</LoaderStyles>);
