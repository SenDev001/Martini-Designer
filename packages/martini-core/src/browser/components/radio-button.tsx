import * as React from "react";
import styled from "styled-components";

export interface RadioButtonProps {
    label?: string;
    name?: string;
    value?: any;
    checked?: boolean;
    onChange?: (value: any) => void;
}

const RadioButtonStyles = styled.div`
    display: grid;
    grid-template-columns: max-content 1fr;
    grid-column-gap: 4px;
    align-items: center;
    margin-bottom: 4px;

    .label {
        cursor: pointer;
    }
`;

export const RadioButton: React.FC<RadioButtonProps> = ({
    label,
    name,
    value,
    checked,
    onChange
}) => {
    const handleChange = () => {
        if (onChange)
            onChange(value);
    };
    const makeButton = () => (
        <input
            type="radio"
            checked={checked}
            name={name}
            onChange={handleChange}
            value={value}
        />
    );

    if (!label)
        return makeButton();

    return <RadioButtonStyles>
        {makeButton()}
        <div className="label" onClick={handleChange}>{label}</div>
    </RadioButtonStyles>;
};
