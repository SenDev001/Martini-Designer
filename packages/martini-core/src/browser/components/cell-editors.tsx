import { MaybePromise } from "@theia/core";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import styled from "styled-components";
import * as yup from "yup";
import { Autocomplete, Suggestion } from "./autocomplete";

export interface CellEditor<T> {
    value: T;
    validate?: (value: T) => string | undefined;
    onApply?: (value: T) => void;
    onCancel?: () => void;
    readOnly: boolean;
}

export const CellEditorProvider = Symbol("CellEditorProvider");

export interface CellEditorProvider<T> {
    (props: CellEditorProps<T>): React.ReactNode;
}

export interface CellEditorProps<T> {
    editor: CellEditor<T>;
}

export interface TextCellEditorProps extends CellEditorProps<string> {
    minLength?: number;
    maxLength?: number;
    suggestions?: Suggestion[];
    autoFocus?: boolean;
}

const TextCellEditorStyles = styled.div`
    width: 100%;
    display: grid;
    margin: 0;
    padding: 0;
    overflow-y: visible;
    position: relative;

    input {
        min-width: 0;
    }
`;

export const TextCellEditor: React.FC<TextCellEditorProps> = ({
    editor,
    minLength,
    maxLength,
    suggestions,
    autoFocus
}) => {
    const schema = React.useMemo(() => yup.string().min(minLength || 0).max(maxLength || Number.MAX_VALUE), []);
    const getValue = () =>
        editor.value === undefined ? "" : editor.value;
    const valueChanged = React.useRef(false);
    const [value, setValue] = React.useState(getValue());
    const [error, setError] = React.useState("");

    const validate = (input: string): string | undefined => {
        try {
            schema.validateSync(input);
        } catch (error) {
            return error.errors[0];
        }

        if (editor.validate)
            return editor.validate(input);
    };

    const onChange = (input: string) => {
        if (editor.readOnly)
            return;
        valueChanged.current = true;
        setValue(input);
        setError(validate(input) || "");
    };

    const apply = (input: string): boolean => {
        if (validate(input) !== undefined) return false;
        if (editor.value !== input && editor.onApply) {
            editor.onApply(input);
            return true;
        }

        if (editor.onCancel)
            editor.onCancel();
        return false;
    };

    const handleKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === "Enter") {
            e.preventDefault();
            e.stopPropagation();
            apply(value);
        }
        else if (e.key === "Escape") {
            e.preventDefault();
            e.stopPropagation();
            setValue(getValue());
            setError("");
            if (editor.onCancel)
                editor.onCancel();
        }
    };

    const handleBlur = (input: string) => {
        if (valueChanged.current) {
            if (!apply(input)) {
                setValue(getValue());
            }
        }
        setError("");
        if (editor.onCancel)
            editor.onCancel();
        valueChanged.current = false;
    };

    React.useEffect(() => setValue(getValue()), [editor.value]);
    React.useEffect(() => setError(""), [editor.value]);

    return (
        <TextCellEditorStyles>
            <Autocomplete
                suggestions={suggestions || []}
                value={value}
                onInputChange={onChange}
                onInputBlur={(_, input) => handleBlur(input)}
                onSuggestionSelect={apply}
                onInputKeyUp={handleKeyUp}
                autoFocus={autoFocus}
                inputStyle={{ margin: 0, outlineColor: error.length !== 0 ? "var(--theia-inputValidation-errorBackground)" : undefined }}
            />
            <ErrorOverlay error={error} />
        </TextCellEditorStyles>
    );
};

export interface NumberCellEditorProps extends CellEditorProps<number> {
    min?: number;
    max?: number;
    allowDecimal?: boolean;
    allowBlank?: boolean;
    suggestions?: number[];
}

export const NumberCellEditor: React.FC<NumberCellEditorProps> = (props) => {
    const schema = React.useMemo(() => {
        let schema = yup.number()
            .min(props.min === undefined ? -Number.MAX_VALUE : props.min)
            .max(props.max === undefined ? Number.MAX_VALUE : props.max);
        if (props.allowDecimal === undefined || !props.allowDecimal)
            schema = schema.integer();
        return schema;
    }, []);
    const editor: CellEditor<string> = {
        value: props.editor.value?.toString(),
        onApply: (value: string) => {
            if (value.length === 0 && props.allowBlank)
                value = "";
            else if (editor.validate!(value) !== undefined)
                return;
            if (props.editor.onApply)
                props.editor.onApply(schema.cast(value)!);
        },
        validate: (value: string) => {
            if (value.length === 0 && props.allowBlank)
                return undefined;
            let _value: number;
            try {
                _value = schema.validateSync(value)!;
            } catch (error) {
                return error.errors[0];
            }
            if (props.editor.validate)
                return props.editor.validate(_value);
        },
        readOnly: props.editor.readOnly
    };
    // @ts-ignore
    return <TextCellEditor {...props}
        editor={editor}
        suggestions={props.suggestions?.map(n => ({ label: n.toString() }))}
    />;
};

export const BooleanCellEditor: React.FC<CellEditorProps<boolean>> = ({
    editor
}) => {
    const getValue = () =>
        editor.value === undefined ? false : editor.value;
    const [value, setValue] = React.useState(getValue());
    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        if (editor.readOnly)
            return;
        setValue(e.target.checked);
        if (editor.onApply)
            editor.onApply(e.target.checked);
    };
    React.useEffect(() => setValue(getValue()), [editor.value]);

    return (
        <input
            type="checkbox"
            checked={value}
            onChange={onChange}
            className="theia-input"
            style={{ marginLeft: "5px" }}
        />
    );
};

export interface ComboCellEditorProps extends CellEditorProps<string> {
    values: ComboCellEditorProps.Option[];
}

export namespace ComboCellEditorProps {
    export interface Option {
        value: any;
        name: string;
    }
}

export const ComboCellEditor: React.FC<ComboCellEditorProps> = ({
    editor,
    values
}) => {
    const getValue = () =>
        editor.value === undefined ? "" : editor.value;
    const [value, setValue] = React.useState(getValue());
    React.useEffect(() => setValue(getValue()), [editor.value]);
    const handleChange = (
        e: React.ChangeEvent<HTMLSelectElement | HTMLInputElement>
    ) => {
        if (editor.readOnly)
            return;
        setValue(e.target.value);
        if (editor.onApply)
            editor.onApply(e.target.value);
    };

    return (
        <select
            value={value}
            onChange={handleChange}
            style={{ width: "100%", margin: 0, padding: 0 }}
            className="theia-input"
        >
            {values.map(value => (
                <option key={value.name} value={value.value}>
                    {value.name}
                </option>
            ))}
        </select>
    );
};

export interface DialogCellEditorProps<T> extends CellEditorProps<T> {
    openDialog: (value: T) => MaybePromise<T>;
    renderLabel?: (value: T) => React.ReactNode;
    tooltip?: string;
}

export const DialogCellEditor: React.FC<DialogCellEditorProps<any>> = ({
    editor,
    openDialog,
    renderLabel,
    tooltip
}) => {
    const getValue = () =>
        editor.value === undefined ? false : editor.value;
    const [value, setValue] = React.useState(getValue());
    React.useEffect(() => setValue(getValue()), [editor.value]);

    const handleClick = () => {
        openDialog(value)?.then((newValue: any) => {
            if (newValue !== undefined && !editor.readOnly) {
                setValue(newValue);
                if (editor.onApply)
                    editor.onApply(newValue);
            }
        });
    };

    const handeKeyDown = (e: React.KeyboardEvent) => {
        if (e.key === "Enter")
            handleClick();
    };

    return (
        <div
            style={{
                display: "grid",
                gridTemplateColumns: "1fr max-content",
                margin: 0,
                padding: "4px",
                width: "100%",
            }}
            title={tooltip || (value instanceof Array ? value.join("\n") : value.toString())}
            tabIndex={0}
            onKeyDown={handeKeyDown}
        >
            <div
                style={{
                    whiteSpace: "nowrap",
                    overflow: "hidden",
                    textOverflow: "ellipsis"
                }}
                onDoubleClick={handleClick}
            >
                {renderLabel && renderLabel(value)}
                {!renderLabel && (value instanceof Array ? value.join(", ") : value)}
            </div>
            <div onClick={handleClick} title={messages.click_to_edit} style={{ cursor: "pointer" }}>
                <div className="martini-tree-icon martini-edit-icon" />
            </div>
        </div>
    );
};


const ErrorOverlayStyles = styled.div`
    position: absolute;
    top: 24px;
    width: 100%;
    background-color: var(--theia-inputValidation-errorBackground);
    z-index: 10;
    text-overflow: ellipsis;
    overflow: hidden;

    .error-message {
        white-space: pre;
        height: 24px;
        font-size: 10px;
        padding: 4px;
        min-width: 0;
        text-overflow: ellipsis;
        overflow: hidden;
    }
`;

export const ErrorOverlay: React.FC<{ error: string; }> = ({ error }) => {
    if (!error || error.length === 0)
        return <></>;

    return <ErrorOverlayStyles>
        <div className="error-message" title={error}>{error.replace(/[\r\n]/g, " ")}</div>
    </ErrorOverlayStyles>;
};
