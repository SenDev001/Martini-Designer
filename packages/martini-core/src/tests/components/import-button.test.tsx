// tslint:disable-next-line:no-var-requires
require("reflect-metadata"); // has to stay at the top
import * as React from "react";
import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import {ImportButton} from "../../browser/components/import-button";
import * as Renderer from "react-test-renderer";

Enzyme.configure({adapter: new Adapter()});

test("ImportButton should be rendered", () => {
    const component = Renderer.create(<ImportButton/>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Changing a input file should call onFileImport", () => {
    let actual: File | undefined;
    const importButton = Enzyme.mount(<ImportButton onFileImport={value => actual = value as File}/>);
    const expected = "test.json";
    const file = new File([], expected);
    const fileList = {
        length: 1,
        item: (_: number) => file
    };
    importButton.find("input").first().simulate("change", {
        target: {
            files: fileList
        }
    });

    expect(actual).toBe(file);
});
