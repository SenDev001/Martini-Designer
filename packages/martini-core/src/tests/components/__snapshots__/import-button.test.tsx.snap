// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportButton should be rendered 1`] = `
<div>
  <input
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <input
    className="theia-button"
    onClick={[Function]}
    style={
      Object {
        "marginLeft": 0,
      }
    }
    type="button"
    value="Import..."
  />
</div>
`;
